### DISCUSSION TITLE
Mini.files - multiple keymapping

### DISCUSSION THREAD

**Discussion by artsbentley** [nvim-mini/mini.nvim#409] [Q&A] [answered]
Hey! Loving the plugin and even though I am really enjoying Oil currently, I really want to switch over to mini.files for various reasons I won't go into detail here.

However, I often switch between a custom configured QMK keyboard and my laptop keyboard, hence I have a strong need to configure keymaps for both options. 

I can't seem to find a way to configure mini.files the same way as I would with my oil.lua config as I will show below, does anyone know of any suitable options for me?

Oil.lua config:

```lua       
   
keymaps = {
                ["?"] = "actions.show_help",

                ["<CR>"] = "actions.select",
                ["<Right>"] = "actions.select",
                ["L"] = "actions.select",

                ["|"] = "actions.select_vsplit",
                ["-"] = "actions.select_split",
                ["<C-t>"] = "actions.select_tab",
                ["p"] = "actions.preview",
                ["q"] = "actions.close",
                ["<Esc>"] = "actions.close",
                ["<C-l>"] = false,

                ["<Bs>"] = "actions.parent",
                ["<Left>"] = "actions.parent",
                ["H"] = "actions.parent",

                ["_"] = "actions.open_cwd",
                ["`"] = "actions.cd",
                ["~"] = "actions.tcd",
                ["."] = "actions.toggle_hidden",
            },
```

**Comment by echasnovski** [comment, marked as answer]
Hi!

> Oil.lua config:
> 
> ```lua
>    
> keymaps = {
>                 ["?"] = "actions.show_help",
> 
>                 ["<CR>"] = "actions.select",
>                 ["<Right>"] = "actions.select",
>                 ["L"] = "actions.select",
> 
>                 ["|"] = "actions.select_vsplit",
>                 ["-"] = "actions.select_split",
>                 ["<C-t>"] = "actions.select_tab",
>                 ["p"] = "actions.preview",
>                 ["q"] = "actions.close",
>                 ["<Esc>"] = "actions.close",
>                 ["<C-l>"] = false,
> 
>                 ["<Bs>"] = "actions.parent",
>                 ["<Left>"] = "actions.parent",
>                 ["H"] = "actions.parent",
> 
>                 ["_"] = "actions.open_cwd",
>                 ["`"] = "actions.cd",
>                 ["~"] = "actions.tcd",
>                 ["."] = "actions.toggle_hidden",
>             },
> ```

Yes, everything listed here seems to be possible, albeit with a slightly different workflow.

General idea is that there are two ways to customize mappings:
- Inside `config.mappings` for a common ones.
- Inside custom autocommand for `MiniFilesBufferCreate` event.

Here is a starting point for you (**Edit**: use new `close_on_file` option in `go_in()`):

```lua
require('mini.files').setup({
  mappings = {
    -- Here 'L' will also close explorer after opening file.
    -- Switch to `go_in` if you want to not close explorer.
    go_in = '',
    go_in_plus = 'L',
    go_out = 'H',
    go_out_plus = '',
    -- Will be overriden by manual `<BS>`, which seems wasteful
    reset = '',
    -- Overrides built-in `?` for backward search
    show_help = '?',
  },

  -- Only automated preview is possible
  windows = { preview = true },
})

local go_in_plus = function()
  for _ = 1, vim.v.count1 do
    MiniFiles.go_in({ close_on_file = true })
  end
end

vim.api.nvim_create_autocmd('User', {
  pattern = 'MiniFilesBufferCreate',
  callback = function(args)
    local map_buf = function(lhs, rhs) vim.keymap.set('n', lhs, rhs, { buffer = args.data.buf_id }) end

    map_buf('<CR>', go_in_plus)
    map_buf('<Right>', go_in_plus)

    map_buf('<BS>', MiniFiles.go_out)
    map_buf('<Left>', MiniFiles.go_out)

    map_buf('<Esc>', MiniFiles.close)

    -- Add extra mappings from *MiniFiles-examples*
  end,
})
```

See [`*MiniFiles-examples*`](https://github.com/echasnovski/mini.nvim/blob/760c1f3619418f769526884a3de47f0b76245887/doc/mini-files.txt#L351) for how to set up "toggle dotfiles", "change target window", and "change current directory".

**Reply by artsbentley** [reply]
AMAZING. Works a charm

thank you so much for all the effort you put into these plugins and the fast and extensive responses to the community. I always see you post very helpful responses on the neovim reddit aswell :)

**Reply by echasnovski** [reply]
@artsbentley, I update the comment to use new `close_on_file` option of `go_in`.

**Comment by 9mm** [comment]
This is great, but is there a way to do it without needing to redefine and hardcode any functions?

Ive tried various versions of this to no avail

```
    config = function()
      local files = require('mini.files')
      files.setup({
        mappings = {
          go_in = '',
          go_in_plus = '<C-l>',
          go_out = '',
          go_out_plus = '<C-h>',
        }
      })

      vim.api.nvim_create_autocmd('User', {
        pattern = 'MiniFilesBufferCreate',
        callback = function(args)
          vim.keymap.set('n', '<CR>', function() files.go_in_plus() end, { buffer = args.data.buf_id })
        end,
      })
    end,
```

**Reply by echasnovski** [reply]
> This is great, but is there a way to do it without needing to redefine and hardcode any functions?

~Not with `go_in_plus()`, no. I can't see a natural way to wrap its full capabilities into `go_in()` itself~.

**Edit**: There is a somewhat straightforward solution now as `go_in()` supports `close_on_file` option. The code below uses it.

Here is the what should work for your case:

```lua
local files = require('mini.files')
files.setup({
  mappings = {
    go_in = '',
    go_in_plus = '<C-l>',
    go_out = '',
    go_out_plus = '<C-h>',
  },
})

local go_in_plus = function()
  for _ = 1, vim.v.count1 do
    files.go_in({ close_on_file = true })
  end
end

vim.api.nvim_create_autocmd('User', {
  pattern = 'MiniFilesBufferCreate',
  callback = function(args) vim.keymap.set('n', '<CR>', go_in_plus, { buffer = args.data.buf_id }) end,
})
```

**Reply by 9mm** [reply]
Ok thank you, I appreciate it

**Reply by bassamsdata** [reply]
I configured multiple keymaps since i installed the plugin when it was in beta. I didn't know this elegant way of doing it (because I'm still learning lua). so unfortunately I did it in a more complex (but logical at the time) way.

```lua
local MiniFiles = require("mini.files")
-- check if it's a directory
local is_dir = function()
  local fs_entry = MiniFiles.get_fs_entry()
  local is_at_dir = fs_entry ~= nil and fs_entry.fs_type == "directory"
  return is_at_dir
end

-- create go_in_plus fn
local open_and_close = function()
  if is_dir() then
    MiniFiles.go_in()
  else
    MiniFiles.go_in()
    MiniFiles.close()
  end
end

vim.api.nvim_create_autocmd("User", {
  pattern = "MiniFilesBufferCreate",
  callback = function()
    vim.keymap.set("n", "<cr>", open_and_close)
  end,
})

```
